Sources/BuildSystemIntegration/BuildSystemManager.swift

```diff
@@ -671,6 +671,26 @@ package actor BuildSystemManager: QueueBasedMessageHandler {
     return languageFromBuildSystem ?? Language(inferredFromFileExtension: document)
   }
 
+  /// Returns the URIs of all source files in the project that have the same realpath as a document in `documents` but
+  /// are not in `documents`.
+  ///
+  /// This is useful in the following scenario: A project has target A containing A.swift an target B containing B.swift
+  /// B.swift is a symlink to A.swift. When A.swift is modified, both the dependencies of A and B need to be marked as
+  /// having an out-of-date preparation status, not just A.
+  package func sourceFilesWithSameRealpath(as documents: [DocumentURI]) async -> [DocumentURI] {
+    let realPaths = Set(documents.map { $0.symlinkTarget ?? $0 })
+    return await orLog("") {
+      var result: [DocumentURI] = []
+      let filesAndDirectories = try await sourceFilesAndDirectories(includeNonBuildableFiles: true)
+      for file in filesAndDirectories.files.keys {
+        if realPaths.contains(file.symlinkTarget ?? file) && !realPaths.contains(file) {
```

Comment:
Nit: Couldn't this be `if let symlinkTarget = file.symlinkTarget, realPaths.contains(symlinkTarget)`?

Reply:
No, because then we would not be reporting an edit in the following scenario:
A is a symlink to B. A and B are in the source files array. A gets modified. We should also consider B as modified but we wouldn’t with your suggested change because `B` doesn’t have a symlink target (since it’s not a symlink).

Reply:
To be clear, my suggestion was equivalent to the original code, since when `symlinkTarget` is `nil` it becomes `realPaths.contains(file) && !realPaths.contains(file)`. That being said, good catch!

Reply:
Ah, and I realized that the second part of the check should have been `!documents.contains(file)` upon revisiting the changes.